### PR TITLE
Fix remapped Cmd+W close shortcuts

### LINKED ISSUE #1488: Cmd+W shortcut cannot be reconfigured in Settings — hardcoded to close tab while all other shortcuts are configurable
## Problem

All keyboard shortcuts in cmux can be reconfigured through the Settings shortcuts UI — **except Cmd+W**. Changing Cmd+W to a different action (e.g. "switch focus") in Settings has no effect. It always closes the current pane/tab regardless of what the user sets.

## Steps to reproduce

1. Open Settings → Shortcuts
2. Change "Close Tab" (Cmd+W) to a different shortcut, or reassign Cmd+W to "Switch Focus" or another action
3. Press Cmd+W in a terminal

## Expected

Cmd+W performs the newly assigned action (or nothing, if unbound).

## Actual

Cmd+W still closes the current pane/tab. The Settings change is ignored.

## Root cause

Cmd+W is hardcoded in `AppDelegate.swift` and `cmuxApp.swift` — it bypasses the `KeyboardShortcutSettings` system that all other shortcuts use:

- `Sources/AppDelegate.swift` hardcodes plain Cmd+W to close the focused panel/tab
- `Sources/cmuxApp.swift` binds `File > Close Tab` to `.keyboardShortcut("w", modifiers: .command)`
- There is no `closeTab` action in `KeyboardShortcutSettings`, so the settings UI cannot override it

## Fix

Add a `closeTab` action to `KeyboardShortcutSettings` and route both the menu binding and the app-level Cmd+W handler through it, the same way all other configurable shortcuts work.

## Related

- [#966](https://github.com/manaflow-ai/cmux/issues/966) — earlier report of the same root cause
- [#1486](https://github.com/manaflow-ai/cmux/issues/1486) — config file keybindings and blank shortcuts

### PR DESCRIPTION
**Pull request by austinywang**
Fixes #1488

## Summary
- Treat stale default close shortcuts as stale even when their key has been reassigned to another configured action.
- Route configured shortcut handling before stale close menu fallback at the application and window key-equivalent boundaries.
- Add regression coverage for Close Tab, Close Workspace, and Close Window defaults, plus Cmd+W reassignment before a stale Close Tab menu item.

## Reproduction / evidence
- Reviewed issue #1488, umbrella #3726, and stale PRs #1490 and #2512.
- Attempted local reproduction against an already-running tagged debug app without launching or rebuilding; that build already suppressed the simple remap-away case, so the regression targets the remaining split-owner fallback path where stale close defaults could still win after reassignment.

## Test structure
- Commit 1 adds the regression tests only.
- Commit 2 adds the fix.

## Testing
- CI: all required checks green on 2026-05-20.
- Tagged dev build: `CMUX_SKIP_ZIG_BUILD=1 ./scripts/reload.sh --tag issue-1488-cmd-w-remap-broken --launch` built successfully locally.
- Launch: `reload.sh` first socket check timed out, then the built tagged app was opened directly; `/tmp/cmux-debug-issue-1488-cmd-w-remap-broken.sock` was created and the app stayed running.

---
<a href="https://app.blacksmith.sh/manaflow-ai/codesmith/cmux/pr/4406"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-light.svg">View in Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need.</sup>

- [ ] Let Codesmith autofix CI failures and bot reviews

---

> [!NOTE]
> **Medium Risk**
> Changes event/shortcut routing for Command-key equivalents (including `Cmd+W`) and stale-menu suppression logic, which can subtly impact global keyboard handling across windows. Added regression tests reduce risk but behavior changes may affect edge-case shortcuts and AppKit menu interactions.
> 
> **Overview**
> Fixes cases where remapped **close** shortcuts (e.g., `Cmd+W`) could still trigger stale AppKit menu items.
> 
> Configured shortcut handling is now dispatched earlier (via new `handleConfiguredShortcutKeyEquivalent`) from both `NSApplication.sendEvent` and window key-equivalent paths, before any stale menu fallback runs.
> 
> Refines `shouldSuppressStaleCmuxMenuShortcut` to (1) treat default close shortcuts as stale even when their key is reassigned, and (2) avoid suppressing events owned by current numbered-digit shortcuts. Adds regression tests covering these scenarios, including an end-to-end `Cmd+W` reassignment against a mocked stale Close Tab menu item.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e9e396cc6121284243c71528f801043b80e88e12. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
## Summary by cubic
Fixes #1488: remapped Cmd+W and other Close shortcuts no longer lose to stale menu items, and numbered-digit shortcuts (e.g., Cmd+2) keep working. Configured shortcuts now run before any menu fallback; Close defaults are considered stale once reassigned.

- **Bug Fixes**
  - Route configured shortcuts before menu fallbacks in `NSApplication.sendEvent` and `NSWindow.performKeyEquivalent` via a shared `handleConfiguredShortcutKeyEquivalent`.
  - Treat `.closeTab`, `.closeWorkspace`, and `.closeWindow` defaults as stale after reassignment; do not suppress current numbered-digit shortcuts.
  - Add isolated regression tests: reassigned Cmd+W beats a stale Close Tab menu via sendEvent; stale Close defaults; numbered-digit shortcuts.

<sup>Written for commit e9e396cc6121284243c71528f801043b80e88e12. Summary will update on new commits. <a href="https://cubic.dev/pr/manaflow-ai/cmux/pull/4406?utm_source=github">Review in cubic</a></sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Configured keyboard shortcuts are now routed before stale menu fallbacks across application and window event paths.
  * Stale menu-shortcut suppression refined with explicit handling for close-related actions and more accurate suppression decisions.

* **Tests**
  * Added regression tests covering reassigned and temporarily remapped shortcuts, ensuring correct routing and suppression behavior for close and numbered-digit actions.

[Review Change Stack](https://app.coderabbit.ai/change-stack/manaflow-ai/cmux/pull/4406?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: Sources/AppDelegate.swift [hunk 1]
```diff
@@ -13526,25 +13534,45 @@ final class AppDelegate: NSObject, NSApplicationDelegate, UNUserNotificationCent
...
     }
 
+    private func isCloseShortcutAction(_ action: KeyboardShortcutSettings.Action) -> Bool {
+        switch action {
+        case .closeTab, .closeWorkspace, .closeWindow:
+            return true
+        default:
+            return false
+        }
+    }
```
[the remark this answers is not in the document]
Comment (austinywang) [reply Sources/AppDelegate.swift:13583]: Reviewed. I am not applying the suggested `@unknown default` because `KeyboardShortcutSettings.Action` is a same-module enum; Swift still emits an exhaustiveness warning for that form unless every non-close case is listed. The helper is intentionally scoped to the three current close actions covered by the regression tests here; a future close-type action should be added to this helper together with its shortcut tests.
